chromadb/quota/__init__.py

```diff
@@ -0,0 +1,60 @@
+from abc import abstractmethod
+from typing import Optional, Literal
+
+from chromadb import Documents, Embeddings
+from chromadb.api import Metadatas
+from chromadb.config import (
+    Component,
+    System,
+)
+Resource = Literal["METADATA_KEY_LENGTH", "METADATA_VALUE_LENGTH", "DOCUMENT_SIZE", "EMBEDDINGS_DIMENSION"]
+
+
+class QuotaError(Exception):
+    def __init__(self, resource: Resource, quota: int, actual: int):
+        super().__init__("Out of quota resource")
+        self.quota = quota
+        self.actual = actual
+        self.resource = resource
+
+
+class QuotaProvider(Component):
+    def __init__(self, system: System) -> None:
+        super().__init__(system)
+        self.system = system
+
+    @abstractmethod
+    def get_for_subject(self, resource: Resource, subject: Optional[str] = None, tier: Optional[str] = None) -> Optional[int]:
+        pass
+
+class QuotaEnforcer(Component):
+    def __init__(self, system: System) -> None:
+        super().__init__(system)
+        self.should_enforce = False
+        if system.settings.chroma_quota_provider_impl:
+            self._quota_provider = system.require(QuotaProvider)
+            self.should_enforce = True
+        self.system = system
+
+    def payload_static_check(self, metadatas: Optional[Metadatas] = None, documents: Optional[Documents] = None, embeddings: Optional[Embeddings]= None, collection_id: Optional[str]= None):
+        if not self.should_enforce:
+            return
+        metadata_key_length_quota = self._quota_provider.get_for_subject(resource="METADATA_KEY_LENGTH", subject=collection_id)
```

Comment:
Does this imply that all checks can only be performed against specific collections? How about quotas at user, tenant, and DB levels? Does this approach force Chroma to maintain per-collection quotas? Why are per-collection quotas more interesting than quotas at other levels of object inheritance?

Reply:
This is just for static one, but yes, later, tenant based quota would also be possible